src/GatewayAPI/Services/CoreNodesSupervisorService.cs

```diff
@@ -0,0 +1,209 @@
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+using Common.Extensions;
+using GatewayAPI.Configuration;
+using GatewayAPI.Configuration.Models;
+using GatewayAPI.CoreCommunications;
+using CoreApiModel = RadixCoreApi.Generated.Model;
+
+namespace GatewayAPI.Services;
+
+public interface ICoreNodesSupervisorService
+{
+    public CoreApiNode GetRandomHealthiestCoreNode();
+
+    public Task ReviseCoreNodesHealth(CancellationToken cancellationToken);
+}
+
+public class CoreNodesSupervisorService : ICoreNodesSupervisorService
+{
+    // Using explicit integers for enum values
+    // because they're used for ordering the nodes (from best to worst).
+    private enum CoreNodeStatus
+    {
+        HealthyAndSynced = 0,
+        Unknown = 1,
+        HealthyButLagging = 2,
+        Unhealthy = 3,
+    }
+
+    private static readonly IEnumerable<CoreNodeStatus> _statusesFromBestToWorst =
+        Enum.GetValues(typeof(CoreNodeStatus)).Cast<CoreNodeStatus>().OrderBy(s => s);
+
+    private readonly ILogger _logger;
+    private readonly HttpClient _httpClient;
+    private readonly CoreApiModel.NetworkIdentifier _networkIdentifier;
+    private readonly long _maxAllowedStateVersionLagToBeConsideredSynced;
+
+    private readonly Dictionary<CoreApiNode, CoreNodeStatus> _coreNodesStatuses;
+    private readonly object _coreNodesStatusesLock = new();
+
+    public CoreNodesSupervisorService(
+        ILogger<CoreNodesSupervisorService> logger,
+        HttpClient httpClient,
+        IGatewayApiConfiguration configuration)
+    {
+        _logger = logger;
+        _httpClient = httpClient;
+
+        _networkIdentifier = new CoreApiModel.NetworkIdentifier(configuration.GetNetworkName());
+        _maxAllowedStateVersionLagToBeConsideredSynced = configuration
+            .GetCoreApiNodeHealth().MaxAllowedStateVersionLagToBeConsideredSynced;
+
+        _coreNodesStatuses = configuration.GetCoreNodes()
+            .Where(n => n.IsEnabled && !string.IsNullOrWhiteSpace(n.CoreApiAddress))
+            .ToDictionary(n => n, _ => CoreNodeStatus.Unknown);
+    }
+
+    public CoreApiNode GetRandomHealthiestCoreNode()
+    {
+        lock (_coreNodesStatusesLock)
```

Comment:
This is on a common path in the request threadpool, and locking feels a little overkill, although probably not a massive deal...

Some alternative ideas in case any of them appeal:
* We put the lock in GetCoreNodesOfStatus instead to minimise the length of time we hold the lock for
* We could use a ConcurrentDictionary
* We precompute the nodes by status into an immutable list, and then just swap out the _nodesByStatus_ reference at update time (reference assignment is atomic). In this method, we just read the nodes by status from the reference and then use the local variable. Or we could even just compute `validNodes` up front - as we know which status we care about till the next cycle.

But also, we should be fine to leave as is I'd wager.

Reply:
My initial approach was with ConcurrentDictionary but it had a (super unlikely) corner case:
Say there is only a single node that's in Lagging status.
We iterate over the statuses in `GetRandomHealthiestCoreNode`, Synced is empty so we proceed to Unknown and then Lagging. Some time in between processing Synced and Lagging another thread updates the statuses in `ReviseCoreNodesHealth` and the node status is changed to Synced. As a result `GetRandomHealthiestCoreNode` reports that there are no available nodes.
For the same reason we can't put the lock inside `GetCoreNodesOfStatus`.

I like the idea of swapping the whole dict :+1: